src/mesh/forest.hpp

```diff
@@ -0,0 +1,346 @@
+      } else {
+        missed = dir;
+      }
+    }
+    if (dirs.size() == 2 && missed == Direction::J) {
+      dirs = {dirs[1], dirs[0]}; // Make sure we are in cyclic order
+    }
+    return dirs;
+  }
+
+  bool IsNode() const {
+    return 3 == abs(static_cast<int>(u[0])) + abs(static_cast<int>(u[1])) +
+                    abs(static_cast<int>(u[2]));
+  }
+
+  bool IsEdge() const {
+    return 2 == abs(static_cast<int>(u[0])) + abs(static_cast<int>(u[1])) +
+                    abs(static_cast<int>(u[2]));
+  }
+
+  bool IsFace() const {
+    return 1 == abs(static_cast<int>(u[0])) + abs(static_cast<int>(u[1])) +
+                    abs(static_cast<int>(u[2]));
+  }
+
+  bool IsCell() const {
+    return 0 == abs(static_cast<int>(u[0])) + abs(static_cast<int>(u[1])) +
+                    abs(static_cast<int>(u[2]));
+  }
+
+  int GetIdx() const {
+    return (static_cast<int>(u[0]) + 1) + 3 * (static_cast<int>(u[1]) + 1) +
+           9 * (static_cast<int>(u[2]) + 1);
+  }
+};
+
+struct RelativeOrientation {
+  RelativeOrientation() : dir_connection{0, 1, 2}, dir_flip{false, false, false} {};
+
+  void SetDirection(Direction origin, Direction neighbor, bool reversed = false) {
+    dir_connection[static_cast<uint>(origin)] = static_cast<uint>(neighbor);
+    dir_flip[static_cast<uint>(origin)] = reversed;
+  }
+
+  LogicalLocation Transform(const LogicalLocation &loc_in,
+                            std::int64_t destination) const;
+  LogicalLocation TransformBack(const LogicalLocation &loc_in, std::int64_t origin) const;
+
+  bool use_offset = false;
+  std::array<int, 3> offset;
+  std::array<int, 3> dir_connection;
+  std::array<bool, 3> dir_flip;
+};
+
+// We don't allow for periodic boundaries, since we can encode periodicity through
+// connectivity in the forest
+class Tree : public std::enable_shared_from_this<Tree> {
+  // This allows us to ensure that Trees are only created as shared_ptrs
+  struct private_t {};
+
+ public:
+  Tree(private_t, std::int64_t id, int ndim, int root_level,
+       RegionSize domain = RegionSize(),
+       std::array<BoundaryFlag, BOUNDARY_NFACES> bcs = {
+           BoundaryFlag::block, BoundaryFlag::block, BoundaryFlag::block,
+           BoundaryFlag::block, BoundaryFlag::block, BoundaryFlag::block});
+
+  template <class... Ts>
+  static std::shared_ptr<Tree> create(Ts &&...args) {
+    auto ptree = std::make_shared<Tree>(private_t(), std::forward<Ts>(args)...);
+    // Make the tree its own central neighbor to reduce code duplication
+    RelativeOrientation orient;
+    orient.use_offset = true;
+    orient.offset = {0, 0, 0};
+    ptree->neighbors[13].insert({ptree, orient});
+    return ptree;
+  }
+
+  // Methods for modifying the tree
+  int AddMeshBlock(const LogicalLocation &loc, bool enforce_proper_nesting = true);
+  int Refine(const LogicalLocation &ref_loc, bool enforce_proper_nesting = true);
+  int Derefine(const LogicalLocation &ref_loc, bool enforce_proper_nesting = true);
```

Comment:
Are we now allowing proper nesting to break?

Reply:
No (at least not for now). These mirror the functionality that was present in `MeshBlockTree` and there were some routines to add blocks that didn't enforce proper nesting. These are used on restart to repopulate a tree that was presumably already properly nested.